lib/IntermediateRepresentation/statement.def

```diff
@@ -0,0 +1,171 @@
+#ifndef HANDLE_STMT
+#define HANDLE_TERM_STMT(num, opcode, Class)
+#else
+#define HANDLE_TERM_STMT(num, opcode, Class) HANDLE_STMT(num, opcode, Class)
+#endif
+#endif
+#ifndef LAST_TERM_STMT
+#define LAST_TERM_STMT(num)
+#endif
+
+#ifndef FIRST_BINARY_STMT
+#define FIRST_BINARY_STMT(num)
+#endif
+#ifndef HANDLE_BINARY_STMT
+#ifndef HANDLE_STMT
+#define HANDLE_BINARY_STMT(num, opcode, instclass)
+#else
+#define HANDLE_BINARY_STMT(num, opcode, Class) HANDLE_STMT(num, opcode, Class)
+#endif
+#endif
+#ifndef LAST_BINARY_STMT
+#define LAST_BINARY_STMT(num)
+#endif
+
+#ifndef FIRST_MEMORY_STMT
+#define FIRST_MEMORY_STMT(num)
+#endif
+#ifndef HANDLE_MEMORY_STMT
+#ifndef HANDLE_STMT
+#define HANDLE_MEMORY_STMT(num, opcode, Class)
+#else
+#define HANDLE_MEMORY_STMT(num, opcode, Class) HANDLE_STMT(num, opcode, Class)
+#endif
+#endif
+#ifndef LAST_MEMORY_STMT
+#define LAST_MEMORY_STMT(num)
+#endif
+
+#ifndef FIRST_CAST_STMT
+#define FIRST_CAST_STMT(num)
+#endif
+#ifndef HANDLE_CAST_STMT
+#ifndef HANDLE_STMT
+#define HANDLE_CAST_STMT(num, opcode, Class)
+#else
+#define HANDLE_CAST_STMT(num, opcode, Class) HANDLE_STMT(num, opcode, Class)
+#endif
+#endif
+#ifndef LAST_CAST_STMT
+#define LAST_CAST_STMT(num)
+#endif
+
+#ifndef FIRST_OTHER_STMT
+#define FIRST_OTHER_STMT(num)
+#endif
+#ifndef HANDLE_LAST_OTHER_STMT
+#ifndef HANDLE_LAST_STMT
+#ifndef HANDLE_OTHER_STMT
+#ifndef HANDLE_STMT
+#define HANDLE_LAST_OTHER_STMT(num, opcode, Class)
+#else
+#define HANDLE_LAST_OTHER_STMT(num, opcode, Class) \
+  HANDLE_STMT(num, opcode, Class)
+#endif
+#else
+#define HANDLE_LAST_OTHER_STMT(num, opcode, Class) \
+  HANDLE_OTHER_STMT(num, opcode, Class)
+#endif
+#else
+#define HANDLE_LAST_OTHER_STMT(num, opcode, Class) \
+  HANDLE_LAST_STMT(num, opcode, Class)
+#endif
+#endif
+#ifndef HANDLE_OTHER_STMT
+#ifndef HANDLE_STMT
+#define HANDLE_OTHER_STMT(num, opcode, Class)
+#else
+#define HANDLE_OTHER_STMT(num, opcode, Class) HANDLE_STMT(num, opcode, Class)
+#endif
+#endif
+#ifndef LAST_OTHER_STMT
+#define LAST_OTHER_STMT(num)
+#endif
+
+// Terminator Instructions - These instructions are used to terminate a basic
+// block of the program.   Every basic block must end with one of these
+// instructions for it to be a well formed basic block.
+FIRST_TERM_STMT(1)
+HANDLE_TERM_STMT(1, Ret,         ReturnStmt)
+HANDLE_TERM_STMT(2, Br,          BranchStmt)
+HANDLE_TERM_STMT(3, Switch,      SwitchStmt)
+HANDLE_TERM_STMT(4, SwitchCase,  SwitchCaseStmt)
+HANDLE_TERM_STMT(5, SwitchType,  SwitchTypeStmt)
+HANDLE_TERM_STMT(6, SwitchRank,  SwitchRankStmt)
+HANDLE_TERM_STMT(7, IndirectBr,  IndirectBrStmt)
+HANDLE_TERM_STMT(8, Unreachable, UnreachableStmt)
+LAST_TERM_STMT(8)
+
+// Standard actions - These instructions capture computations as
+// evalute::expressions
+FIRST_BINARY_STMT(9)
+HANDLE_BINARY_STMT(9,  Assign,    AssignmentStmt)
+HANDLE_BINARY_STMT(10, PtrAssign, PointerAssignStmt)
+HANDLE_BINARY_STMT(11, LblAssign, LabelAssignStmt)
+HANDLE_BINARY_STMT(12, Expr,      ExprStmt)
```

Comment:
Fortran, unlike C, doesn't have expression statements.  Or is this for capturing values into temporaries?

Reply:
The latter.  I'll come up with a new name.